File: src/main/java/com/cfgglobal/test/util/aws/AmazonService.java
package com.cfgglobal.test.util.aws;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.*;
import com.cfgglobal.test.config.aws.AmazonProperties;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.stereotype.Service;

import java.io.*;

@Slf4j
@Service
@EnableConfigurationProperties(AmazonProperties.class)
public class AmazonService {

    @Autowired
    private AmazonProperties aws;

    @Autowired
    private AmazonS3 s3Client;

    /**
     * @param data to be uploaded
     * @return the unique key start the file in AWS S3
     */
    public String upload(UploadData data) {
        String key = aws.getKey() + data.getName();
        ObjectMetadata metadata = new ObjectMetadata();
        metadata.setContentLength(data.getData().length);
        PutObjectResult result = s3Client.putObject(new PutObjectRequest(aws.getBucketName(), key, new ByteArrayInputStream(data.getData()), metadata));
        log.info("file uploaded result: {}", result);
        return aws.getRegionLink() + key;
    }

    public byte[] download(String fileName) {
        String key = aws.getKey() + fileName;
        log.info("Downloading from S3: {}" + key);
        S3Object object = s3Client.getObject(new GetObjectRequest(aws.getBucketName(), key));
        byte[] data = null;
        try (InputStream reader = object.getObjectContent()) {
            data = IOUtils.toByteArray(reader);
        } catch (IOException e) {
            log.error("Error in downloading object from S3", e);
        }
        return data;
    }

    @SneakyThrows
    public InputStream downloadInputStream(String fileName) {
        String key = aws.getKey() + fileName;
        log.info("Downloading from S3: {}" + key);
        S3Object object = s3Client.getObject(new GetObjectRequest(aws.getBucketName(), key));
        return object.getObjectContent();
    }

    @SneakyThrows
    public File getFile(String fileName) {
        String key = aws.getKey() + fileName;
        log.info("Downloading from S3: {}" + key);
        S3Object object = s3Client.getObject(new GetObjectRequest(aws.getBucketName(), key));
        InputStream inputStream = object.getObjectContent();


        File targetFile = new File("/tmp/" + fileName);
        OutputStream outStream = new FileOutputStream(targetFile);
        IOUtils.copy(inputStream, outStream);
        return targetFile;
    }

    public void remove(String fileName) {
        String key = aws.getKey() + fileName;
        s3Client.deleteObject(new DeleteObjectRequest(aws.getBucketName(), key));
    }

    /**
     * Get the url for Public access.
     * The permission need to set to public on s3 service.
     *
     * @param key
     * @return
     */
    public String getUrl(String key) {
        return s3Client.getUrl(aws.getBucketName(), key).toString();
    }

}

File: src/main/kotlin/com/github/leon/aws/s3/AmazonService.kt
package com.github.leon.aws.s3

import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.model.*
import com.github.leon.ams.s3.AmazonProperties
import com.github.leon.ams.s3.UploadData
import lombok.SneakyThrows
import lombok.extern.slf4j.Slf4j
import org.apache.commons.io.IOUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.stereotype.Service

import java.io.*

@Slf4j
@Service
@EnableConfigurationProperties(AmazonProperties::class)
class AmazonService {

    @Autowired
    private val aws: AmazonProperties? = null

    @Autowired
    private val s3Client: AmazonS3? = null

    /**
     * @param data to be uploaded
     * @return the unique key start the file in AWS S3
     */
    fun upload(data: UploadData): String {
        val key = aws!!.key + data.name
        val metadata = ObjectMetadata()
        metadata.contentLength = data.data.size.toLong()
        val result = s3Client!!.putObject(PutObjectRequest(aws.bucketName, key, ByteArrayInputStream(data.data), metadata))
        log.info("file uploaded result: {}", result)
        return aws.regionLink + key
    }

    fun download(fileName: String): ByteArray? {
        val key = aws!!.key + fileName
        log.info("Downloading from S3: {}$key")
        val `object` = s3Client!!.getObject(GetObjectRequest(aws.bucketName, key))
        var data: ByteArray? = null
        try {
            `object`.objectContent.use { reader -> data = IOUtils.toByteArray(reader) }
        } catch (e: IOException) {
            log.error("Error in downloading object from S3", e)
        }

        return data
    }

    @SneakyThrows
    fun downloadInputStream(fileName: String): InputStream {
        val key = aws!!.key + fileName
        log.info("Downloading from S3: {}$key")
        val `object` = s3Client!!.getObject(GetObjectRequest(aws.bucketName, key))
        return `object`.objectContent
    }

    @SneakyThrows
    fun getFile(fileName: String): File {
        val key = aws!!.key + fileName
        log.info("Downloading from S3: {}$key")
        val `object` = s3Client!!.getObject(GetObjectRequest(aws.bucketName, key))
        val inputStream = `object`.objectContent


        val targetFile = File("/tmp/$fileName")
        val outStream = FileOutputStream(targetFile)
        IOUtils.copy(inputStream, outStream)
        return targetFile
    }

    fun remove(fileName: String) {
        val key = aws!!.key + fileName
        s3Client!!.deleteObject(DeleteObjectRequest(aws.bucketName, key))
    }

    /**
     * Get the url for Public access.
     * The permission need to set to public on s3 service.
     *
     * @param key
     * @return
     */
    fun getUrl(key: String): String {
        return s3Client!!.getUrl(aws!!.bucketName, key).toString()
    }

}
